### ISSUE TITLE
[Bug] python code will be hung in beeline

### ISSUE THREAD

**Issue by yabola** [apache/kyuubi#5142]
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

simple example:
`
SET kyuubi.operation.language=python;
import pandas as pd
from pyspark.sql.functions import pandas_udf
spark.createDataFrame(pd.DataFrame([1, 2, 3], columns=["v"])).collect();
`
The pyspark task will get stuck.
<img width="1791" alt="image" src="https://github.com/apache/kyuubi/assets/31469905/54f088fc-1ee1-4b1f-8772-6b8880fc0c2b">

thread dump in executor:
<img width="1547" alt="image" src="https://github.com/apache/kyuubi/assets/31469905/8d2489b7-86a3-4609-bb76-aa99a0b33943">

### Affects Version(s)

master

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to fix.
- [ ] No. I cannot submit a PR at this time.

**Comment by yabola** [comment]
@cfmcgrady Sorry to bother you, I don't know if you have encountered this problem, but I think it is a spark problem.

**Comment by cfmcgrady** [comment]
> @cfmcgrady Sorry to bother you, I don't know if you have encountered this problem, but I think it is a spark problem.

In which mode is Spark deployed, and could you kindly provide additional logs (driver/executor) for reference, if available?

**Comment by yabola** [comment]
@cfmcgrady Thank you for your help. I'm also working on this problem, try to change to a spark version. My spark version is apche spark-3.3.2-bin-hadoop2.
My spark driver log:
[kyuubi-spark-sql-engine.txt](https://github.com/apache/kyuubi/files/12286859/kyuubi-spark-sql-engine.txt)
executor log:
[executor.txt](https://github.com/apache/kyuubi/files/12286862/executor.txt)

**Comment by cfmcgrady** [comment]
do you correctly set up the Python environments on the executor?

FYI:

Python Package Management

- https://spark.apache.org/docs/latest/api/python/user_guide/python_packaging.html
- https://github.com/apache/kyuubi/pull/3899

**Comment by yabola** [comment]
@cfmcgrady Yeah.. I can submit the same pySpark job on yarn but will be stuck on kyuubi. 
In addition, I found that kyuubi on spark 3.2 will not be stuck but will be stuck on spark-3.3.2-bin-hadoop2 or higher ( I also test on the latest spark master branch).
It will be stuck on `org.apache.spark.api.python.PythonRunner#newReaderIterator` 
https://github.com/apache/spark/blob/5c94565e75cc32e552414f27da8082591121a86d/core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala#L717-L753

**Comment by cfmcgrady** [comment]
@yabola it only happened on the YARN mode?

**Comment by yabola** [comment]
@cfmcgrady I test on yarn environment( but I only have one sandbox node , standalone ...) and local mode. spark local mode can work but yarn client mode will be stuck.

**Comment by yabola** [comment]
@cfmcgrady I know the reason, because I `pip install pyspark` before, resulting in the version of pyspark 3.2 in the python environment variable. But I use spark 3.3 and  the implementation of `worker.py` has been modified, causing the spark `PythonRunner` to fail to `readint` forever( like a deadlock, `worker.py` waits for `PythonRunner` to write data, and pythonrunner waits for `worker.py` to write data)

**Comment by yabola** [comment]
@cfmcgrady I know the reason for the blocking, because I `pip install pyspark` according to the document before, so I installed 3.2.2 pyspark into the python3 environment. when I use spark3.3, the environment variable transfer of kyuubi may be different from spark not the same, causing spark to use 3.2 pyspark ( reading and writing of worker.py and 3.3 worker.py are different, which leads to the interlocking of communication with spark worker jvm socket)
